modulus/models/afno/modafno.py

```diff
@@ -0,0 +1,649 @@
+# SPDX-FileCopyrightText: Copyright (c) 2023 - 2024 NVIDIA CORPORATION & AFFILIATES.
+# SPDX-FileCopyrightText: All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from dataclasses import dataclass
+from functools import partial
+from typing import List, Type, Union
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+import modulus  # noqa: F401 for docs
+import modulus.models.layers.fft as fft
+from modulus.models.afno.afno import AFNO2DLayer, AFNOMlp, PatchEmbed
+from modulus.models.meta import ModelMetaData
+from modulus.models.module import Module
+
+from .modembed import ModEmbedNet
+
+Tensor = torch.Tensor
+
+
+class ScaleShiftMlp(nn.Module):
+    """MLP used to compute the scale and shift parameters of the ModAFNO block
+
+    Parameters
+    ----------
+    in_features : int
+        Input feature size
+    out_features : int
+        Output feature size
+    hidden_features : int, optional
+        Hidden feature size, defaults to 2 * out_features
+    hidden_layers : int, optional
+        Number of hidden layers, defaults to 0
+    activation_fn : nn.Module, optional
+        Activation function, by default nn.GELU
+    """
+
+    def __init__(
+        self,
+        in_features: int,
+        out_features: int,
+        hidden_features: Union[int, None] = None,
+        hidden_layers: int = 0,
+        activation_fn: Type[nn.Module] = nn.GELU,
+    ):
+        super().__init__()
+        if hidden_features is None:
+            hidden_features = out_features * 2
+
+        sequence = [nn.Linear(in_features, hidden_features), activation_fn()]
+        for _ in range(hidden_layers):
+            sequence += [nn.Linear(hidden_features, hidden_features), activation_fn()]
+        sequence.append(nn.Linear(hidden_features, out_features * 2))
+        self.net = nn.Sequential(*sequence)
+
+    def forward(self, x: Tensor):
+        (scale, shift) = torch.chunk(self.net(x), 2, dim=1)
+        return (1 + scale, shift)
```

Comment:
Why a plus 1 here?

Reply:
The net gives pretty small values after initialization so with the +1 we get scale close to 1 and shift close to 0 in the beginning of training. This turned out to lead to more stable training.

Reply:
Makes sense, just curious :+1: